content/blog/towards-continuous-performance-regression-testing.asciidoc

```diff
@@ -0,0 +1,594 @@
+====
+
+To emphasize the key part again, this allocation is per _request_, it is independent from wall clock time and thus is not dependent on the machine running the test
+(i.e. the test should behave the same when running on a developer laptop and on a CI machine),
+nor is it subject to volatility induced by other workloads running concurrently.
+
+Based on that, the actual test could look like so:
+
+[source, java]
+----
+@Test
+@EnableEvent("jdk.ObjectAllocationInNewTLAB")
+@EnableEvent("jdk.ObjectAllocationOutsideTLAB")
+public void retrieveTodo() throws Exception {
+  Random r = new Random();
+  HttpClient client = HttpClient.newBuilder().build();
+
+  for (int i = 1; i<= 20_000; i++) { // <1>
+    executeRequest(r, client);
+  }
+
+  jfrEvents.awaitEvents();
+  jfrEvents.reset();
+
+  for (int i = 1; i<= 10_000; i++) { // <2>
+    executeRequest(r, client);
+  }
+
+  jfrEvents.awaitEvents();
+
+  long sum = jfrEvents.filter(this::isObjectAllocationEvent)
+      .filter(this::isRelevantThread)
+      .mapToLong(this::getAllocationSize)
+      .sum();
+
+  assertThat(sum / 10_000).isLessThan(33_000); // <3>
+}
+----
+<1> Warm-up phase
+<2> The actual test phase
+<3> Assert the memory allocation per request is within the expected boundary; note we could also add a _lower_ boundary,
+so to make sure we notice any future improvements (e.g. caused by upgrading to new efficient versions of a library),
+which otherwise may hide subsequent regressions
+
+Now let's assume we've wrapped up the initial round of work on this application, and its tests have been passing on CI for a while.
+One day, the `retrieveTodo()` performance test method fails though:
+
+[source]
+----
+java.lang.AssertionError: 
+Expecting:
+ <388370L>
+to be less than:
+ <33000L> 
+----
+
+Ugh, it's suddenly allocating about ten times more memory per request than before!
+What has happened?
+To find the answer, we can take a look at the test's JFR recording, which JfrUnit persists under _target/jfrunit_:
+
+[source,bash]
+----
+ls target/jfrunit
+
+dev.morling.demos.quarkus.TodoResourcePerformanceTest-createTodo.jfr
+dev.morling.demos.quarkus.TodoResourcePerformanceTest-retrieveTodo.jfr
+----
+
+Let's open the *.jfr file for the failing test in JDK Mission Control (JMC) in order to analyse all the recorded events
+(note that the recording will always contain some JfrUnit-internal events which are needed for synchronizing the recording stream and the events exposed to the test).
+
+When taking a look at the TLAB events of the application's executor thread,
+the culprit is identified quickly;
+a lot of the sampled TLAB allocations contain this stack trace:
+
+image::continuous_perf_testing_tlab_in_jmc.png[TLAB allocations in JDK Mission Control]
+
+Interesting, REST Assured loading a Jackson object mapper, what's going on there?
+Here's the full stacktrace:
+
+image::continuous_perf_testing_tlab_stacktrace.png[Complete stacktrace of the TLAB allocation]
```

Comment:
hugh, this really need to allow zooming in

Reply:
Yes, working on making the images click-zoomable.